src/formats/TexFile.cpp

```diff
@@ -1,191 +1,191 @@
+    quint32 paletteSectionSize = header.nbPalettes > 0 ? header.paletteSize * 4 : 0;
+    quint32 imageSectionSize = w * h * header.bytesPerPixel;
+    quint32 colorKeySectionSize = header.hasColorKeyArray ? header.nbPalettes : 0;
+
+    if ((quint32)data.size() != headerSize + paletteSectionSize + imageSectionSize + colorKeySectionSize) {
+        qWarning() << "tex invalid size!" << data.size() << headerSize << paletteSectionSize << imageSectionSize
+                   << colorKeySectionSize;
+        return false;
+    }
+
+    quint32 i;
+
+    if (header.nbPalettes > 0) {
+        quint32 imageStart = headerSize + paletteSectionSize;
+        for (quint32 palID=0; palID < header.nbPalettes; ++palID) {
+            quint32 paletteStart = headerSize+header.nbColorsPerPalette1*4*palID;
+            _image = QImage(w, h, QImage::Format_Indexed8);
+            QList<QRgb> colors;
+            for (i=0; i<header.nbColorsPerPalette1; ++i) {
+                quint32 index = paletteStart + i*4;
+                colors.append(qRgba(data.at(index+2), data.at(index+1), data.at(index), data.at(index+3)));
+            }
+            _colorTables.append(colors);
+        }
+        _image.setColorTable(_colorTables.first());
+        for (i=0; i<imageSectionSize; ++i)
+            _image.setPixel(i % w, i / w, (quint8)data.at(imageStart+i));
+
+        if (header.hasColorKeyArray) {
+            quint32 colorKeyStart = imageStart + imageSectionSize;
+            for (quint32 j=0; j<header.nbPalettes; ++j)
+                colorKeyArray.append(data.at(colorKeyStart+j));
+        }
+    } else {
+        quint16 color;
+        _image = QImage(w, h, QImage::Format_ARGB32);
+        QRgb *pixels = (QRgb *)_image.bits();
+        if (header.bytesPerPixel < 2 || header.bytesPerPixel > 4) {
+            qWarning() << "tex invalid bytesPerPixel!" << header.bytesPerPixel;
+            return false;
+        }
+        for (i=0; i<imageSectionSize; i+=header.bytesPerPixel) {
+            if (header.bytesPerPixel == 2) {
+                memcpy(&color, &constData[headerSize+i], 2);
+                pixels[i/2] = PsColor::fromPsColor(color);
+            } else if (header.bytesPerPixel == 3) {
+                pixels[i/3] = qRgb(constData[headerSize+i], constData[headerSize+i+1], constData[headerSize+i+2]);
+            } else if (header.bytesPerPixel == 4) {
+                pixels[i/4] = qRgba(constData[headerSize+i], constData[headerSize+i+1], constData[headerSize+i+2], constData[headerSize+i+3]);
+            }
+        }
+    }
+    return true;
+}
+
+void TexFile::setVersion(Version version)
+{
+    header.version = int(version);
+}
+
+void TexFile::updateHeader()
+{
+    header.nbPalettes = _colorTables.size();
+    header.hasPal = !_colorTables.isEmpty();
+    header.imageWidth = _image.width();
+    header.imageHeight = _image.height();
+    header.hasColorKeyArray = !colorKeyArray.isEmpty();
+}
+
+bool TexFile::save(QByteArray &data)
```

Comment:
Check Saving is working same as before for both paletted and non paletted files.